tracer/src/Datadog.Trace/Debugger/Instrumentation/AsyncMethodDebuggerInvoker.cs

```diff
@@ -0,0 +1,367 @@
+        /// </summary>
+        /// <param name="kickoffInfo">The async kickoff method info</param>
+        /// <returns>Live debugger state</returns>
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        internal static AsyncMethodDebuggerState SetContext(AsyncHelper.AsyncKickoffMethodInfo kickoffInfo)
+        {
+            var currentState = AsyncContext.Value;
+            var newState = new AsyncMethodDebuggerState
+            {
+                Parent = currentState,
+                KickoffInvocationTarget = kickoffInfo.KickoffParentObject,
+                KickoffInvocationTargetType = kickoffInfo.KickoffParentType,
+                KickoffMethod = kickoffInfo.KickoffMethod
+            };
+            AsyncContext.Value = newState;
+            return newState;
+        }
+
+        /// <summary>
+        /// Restore the async context to the parent
+        /// </summary>
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        internal static void RestoreContext()
+        {
+            AsyncContext.Value = AsyncContext.Value.Parent;
+        }
+
+        /// <summary>
+        /// Begin Method Invoker
+        /// </summary>
+        /// <typeparam name="TTarget">Target object of the method. Note that it could be typeof(object) and not a concrete type</typeparam>
+        /// <param name="probeId">The id of the probe</param>
+        /// <param name="instance">Instance value</param>
+        /// <param name="methodHandle">The handle of the executing method</param>
+        /// <param name="typeHandle">The handle of the type</param>
+        /// <param name="methodMetadataIndex">The index used to lookup for the <see cref="MethodMetadataInfo"/> associated with the executing method</param>
+        /// <param name="isReEntryToMoveNext">If it the first entry to the state machine MoveNext method</param>
+        /// <returns>Live debugger state</returns>
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        public static AsyncMethodDebuggerState BeginMethod<TTarget>(string probeId, TTarget instance, RuntimeMethodHandle methodHandle, RuntimeTypeHandle typeHandle, int methodMetadataIndex, ref bool isReEntryToMoveNext)
+        {
+            if (ProbeRateLimiter.Instance.IsLimitReached)
+            {
+                return AsyncMethodDebuggerState.CreateInvalidatedDebuggerState();
+            }
+
+            if (isReEntryToMoveNext)
+            {
+                return AsyncContext.Value;
+            }
+
+            isReEntryToMoveNext = true;
+            var kickoffInfo = AsyncHelper.GetAsyncKickoffMethodInfo(instance);
+            if (kickoffInfo.KickoffParentObject == null && kickoffInfo.KickoffMethod.IsStatic == false)
+            {
+                Log.Error($"{nameof(BeginMethod)}: hoisted 'this' has not found. {kickoffInfo.KickoffParentType.Name}.{kickoffInfo.KickoffMethod.Name}");
+            }
+
+            var asyncState = SetContext(kickoffInfo);
```

Comment:
I'd prefer placing it right after the following if statement for the good order.
And consider handing over to the `SetContext` the following block of setters:
https://github.com/DataDog/dd-trace-dotnet/blob/34efd4c24739a4f208894863af5ce022ed084ea0/tracer/src/Datadog.Trace/Debugger/Instrumentation/AsyncMethodDebuggerInvoker.cs#L118-L121

Reply:
Fixed in 871a652d9b150d63443d87a542e34d4643145adf